client/src/main/scala/io/chrisdavenport/circuit/http4s/client/CircuitedClient.scala

```diff
@@ -1,13 +1,138 @@
+    generic(Function.const[CircuitBreaker[F], Request[F]](cr))(c)
+
+
+  def byKey[F[_]](
+    maxFailures: Int,
+    resetTimeout: FiniteDuration,
+    exponentialBackoffFactor: Double = 1,
+    maxResetTimeout: Duration = Duration.Inf,
+    modifications: CircuitBreaker[F] => CircuitBreaker[F] = {x: CircuitBreaker[F] => x},
+    shouldFail: (Request[F], Response[F]) => Boolean = defaultShouldFail[F](_, _)
+  )(client: Client[F])(implicit F: Sync[F], C: Clock[F]): F[Client[F]] = {
+    io.chrisdavenport.mapref.MapRef.ofSingleImmutableMap[F, RequestKey, CircuitBreaker.State](Map.empty[RequestKey, CircuitBreaker.State]).map{
+      mapref => 
+        val f : Request[F] => CircuitBreaker[F] = {req: Request[F] => 
+          val key = RequestKey.fromRequest(req)
+          val optRef = mapref(key)
+          val ref = new LiftedRefDefaultStorage(optRef, CircuitBreaker.Closed(0))
+          val cbInit = CircuitBreaker.unsafe(ref, maxFailures, resetTimeout, exponentialBackoffFactor, maxResetTimeout, F.unit, F.unit, F.unit, F.unit)
+          modifications(cbInit)
+        }
+        generic(f, shouldFail)(client)
+    }
+  }
+
+  def generic[F[_], A](
+    cbf: Request[F] => CircuitBreaker[F],
+    shouldFail: (Request[F], Response[F]) => Boolean = defaultShouldFail[F](_, _)
+  )(client: Client[F])(implicit F: Bracket[F, Throwable]): Client[F] = {
+    Client[F]{req: Request[F] => 
+      val cb = cbf(req)
+      Resource(
+        cb.protect(
+          client.run(req).allocated.flatMap{
+            case (resp, shutdown) => 
+              if (shouldFail(req, resp)) F.raiseError[(Response[F], F[Unit])](CircuitedClientThrowable(resp, shutdown))
+              else (resp, shutdown).pure[F]
+          }
+        ).handleErrorWith{
+          // The F here is not checked.
+          case e: CircuitedClientThrowable[F] @unchecked => (e.resp, e.shutdown).pure[F]
+          case e => F.raiseError(e)
+        }
+      )
+      
+    }
+  }
+
+  def defaultShouldFail[F[_]](req: Request[F], resp: Response[F]): Boolean = {
+    val _ = req
+    if (resp.status.code >= 500) true
+    else false
+  }
+
+  private case class CircuitedClientThrowable[F[_]](resp: Response[F], shutdown: F[Unit]) 
+    extends Throwable 
+    with scala.util.control.NoStackTrace
+
+
+
+  
+  
+  implicit private val eqState: Eq[CircuitBreaker.State] = Eq.instance{
+    case (CircuitBreaker.HalfOpen, CircuitBreaker.HalfOpen) => true
+    case (CircuitBreaker.Closed(i), CircuitBreaker.Closed(i2)) => i === i2
+    case (CircuitBreaker.Open(started1, reset1), CircuitBreaker.Open(started2, reset2)) => 
+      started1 === started2 &&
+        reset1 === reset2
+    case (_, _) => false
+  }
+
+
+  /**
+   * Operates with default and anytime default is present instead information is removed from underlying ref.
+   **/
+  private class LiftedRefDefaultStorage[F[_]: Functor, A: Eq](
+    val ref: Ref[F, Option[A]],
+    val default: A
+  ) extends Ref[F, A]{
```

Comment:
Does this have to be a full `Ref` to participate in the circuit breaker library?  I'm very surprised this isn't a sealed type.

Reply:
Yes. But Ref is an interface and this will pass all tests. I also use this feature to make the ref we use in mapref so I'm very glad that it is not sealed.